File: Task2/apz-pzpi-21-11-milennyi-ivan-task2/src/db/services/feeding_log_service.rs
use std::sync::Arc;
use async_trait::async_trait;
use sqlx::{MySql, Pool, query, query_as};
use crate::db::service_error::ServiceError;
use crate::db::traits::{FeedingLogManage, Service};
use crate::models::FeedingLog;
use crate::view_models::FeedingLogVM;

pub(crate) struct FeedingLogService<T> {
    pool: Arc<T>,
}
#[async_trait]
impl Service<Pool<MySql>> for FeedingLogService<Pool<MySql>> {
    type Model = FeedingLog;
    type Error = ServiceError;
    type ViewModel = FeedingLogVM;
    fn new(pool: Arc<Pool<MySql>>) -> Self {
        FeedingLogService { pool }
    }

    async fn create(&self, item: Self::Model) -> Result<Self::Model, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            INSERT INTO FeedingLogs (sheep_id, shepherd_id, timestamp, feed_id, amount)
            VALUES (?, ?, ?, ?)
            RETURNING id, sheep_id, shepherd_id, timestamp, feed_id, amount
            "#
        )
        .bind(item.sheep_id())
        .bind(item.shepherd_id().ok_or(ServiceError::CustomError("ID is required".to_string()))?)
        .bind(item.timestamp())
        .bind(item.feed_id())
        .bind(item.amount())
        .fetch_one(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn delete(&self, item_id: u64) -> Result<(), Self::Error> {
        query(
            r#"
            DELETE FROM FeedingLogs
            WHERE id = ?
            "#
        )
        .bind(item_id)
        .execute(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
        .map(|result|
            if result.rows_affected() == 0 {
                Err(ServiceError::CustomError("Zero rows affected".to_string()))
            }
            else{
                Ok(())
            }
        )
        .unwrap_or_else(|error| Err(error))
    }

    async fn update(&self, _item: Self::Model) -> Result<Self::Model, Self::Error> {
        Err(ServiceError::ForbiddenError)
    }

    async fn get_all(&self) -> Result<Vec<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM FeedingLogs
            "#
        )
        .fetch_all(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn get_by_id(&self, id: u64) -> Result<Option<Self::Model>, Self::Error> {
        query_as::<_, Self::Model>(
            r#"
            SELECT * FROM FeedingLogs
            WHERE id = ?
            "#
        )
        .bind(id)
        .fetch_optional(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }
}
#[async_trait]
impl FeedingLogManage<Pool<MySql>> for FeedingLogService<Pool<MySql>>{
    async fn get_all_vms_by_sheep_id(&self, id: u64) -> Result<Vec<Self::ViewModel>, Self::Error> {
        query_as::<_, FeedingLogVM>(
            r#"
            SELECT
            fl.id,
            fl.timestamp,
            fl.amount,
            sh.name AS shepherd_name,
            sh.surname AS shepherd_surname,
            fl.sheep_id
            FROM FeedingLogs fl
            LEFT JOIN Shepherds sh ON fl.shepherd_id = sh.id
            WHERE fl.sheep_id = ?
            "#
        )
        .bind(id)
        .fetch_all(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }

    async fn get_all_vms_by_feed_id(&self, id: u64) -> Result<Vec<Self::ViewModel>, Self::Error> {
        query_as::<_, FeedingLogVM>(
            r#"
            SELECT
            fl.id,
            fl.timestamp,
            fl.amount,
            sh.name AS shepherd_name,
            sh.surname AS shepherd_surname,
            fl.sheep_id
            FROM FeedingLogs fl
            LEFT JOIN Shepherds sh ON fl.shepherd_id = sh.id
            WHERE fl.feed_id = ?
            "#
        )
        .bind(id)
        .fetch_all(&*self.pool).await
        .map_err(|error| ServiceError::DatabaseError(error))
    }
}

File: Task4/apz-pzpi-21-11-milennyi-ivan-task4/src/api/services/FeedingLogService.ts
import axios from "axios";
import { FeedingLog } from "../../models/FeedingLog";
import { API_URL } from "../../utils/config";
import { FeedingLogVM } from "../../viewModels/FeedingLogVM";
import { IFeedingLogService } from "../interfaces/IFeedingLogService";

export class FeedingLogService implements IFeedingLogService {
    private static FEEDING_LOG_URLS = {
        GET_ALL: `${API_URL}/feeding-log`,
        GET_BY_ID: (id: number) => `${API_URL}/feeding-log/${id}`,
        CREATE: `${API_URL}/feeding-log/create`,
        DELETE: (id: number) => `${API_URL}/feeding-log/delete/${id}`,
        GET_ALL_VMS_BY_SHEEP_ID: (id: number) =>  `${API_URL}/feeding-log/sheep/${id}`,
        GET_ALL_VMS_BY_FEED_ID: (id: number) =>  `${API_URL}/feeding-log/feed/${id}`
    }
    async create(item: FeedingLog): Promise<FeedingLog> {
        const response = await axios.post<FeedingLog>(FeedingLogService.FEEDING_LOG_URLS.CREATE, item);
        return response.data;
    }
    async delete(itemId: number): Promise<void> {
        await axios.delete(FeedingLogService.FEEDING_LOG_URLS.DELETE(itemId));
    }
    async update(_item: FeedingLog): Promise<FeedingLog> {
        throw new Error("Method is forbidden.");
    }
    async getAll(): Promise<FeedingLog[]> {
        const response = await axios.get<FeedingLog[]>(FeedingLogService.FEEDING_LOG_URLS.GET_ALL);
        return response.data;
    }
    async getById(id: number): Promise<FeedingLog | null> {
        const response = await axios.get<FeedingLog | null>(FeedingLogService.FEEDING_LOG_URLS.GET_BY_ID(id));
        return response.data;
    }
    async getAllVMsBySheepId(id: number): Promise<FeedingLogVM[]> {
        const response = await axios.get<[]>(FeedingLogService.FEEDING_LOG_URLS.GET_ALL_VMS_BY_SHEEP_ID(id));
        const vms: FeedingLogVM[] = [];
        response.data.map(feedingLog => {
            //@ts-ignore
            vms.push(new FeedingLogVM(feedingLog.id, feedingLog.timestamp, feedingLog.amount, feedingLog.sheep_id, feedingLog.shepherd_name, feedingLog.shepherd_surname));
        });
        return vms;
    }
    async getAllVMsByFeedId(id: number): Promise<FeedingLogVM[]> {
        const response = await axios.get<[]>(FeedingLogService.FEEDING_LOG_URLS.GET_ALL_VMS_BY_FEED_ID(id));
        const vms: FeedingLogVM[] = [];
        response.data.map(feedingLog => {
            //@ts-ignore
            vms.push(new FeedingLogVM(feedingLog.id, feedingLog.timestamp, feedingLog.amount, feedingLog.sheep_id, feedingLog.shepherd_name, feedingLog.shepherd_surname));
        });
        return vms;
    }
}